### DISCUSSION TITLE
Add UPnP support

### DISCUSSION THREAD

**Discussion by devnoname120** [OpenRCT2/OpenRCT2#15663] [Feature requests]
Most IPv4 routers don't forward requests made through unconventional ports. Fortunately there exists a protocol that is supported by a fair part of them, and which allows to open ports on demand: UPnP.

Would it be possible to implement it in OpenRCT2?

One hassle-free portable library for using it is [MiniUPnP](https://github.com/miniupnp/miniupnp)

**Comment by janisozaur** [comment]
I've started doing already, at least some testing required for that.

**Comment by devnoname120** [comment]
Nice! I hope that you use MiniUPnP. From my experience this is the most reliable UPnP library.

**Comment by janisozaur** [comment]
Yeah, I tested with `libnatpmp`, to be precise, the library component of miniupnp for supporting IGD and intend to go with that.

What's your system @devnoname120?

**Comment by devnoname120** [comment]
Currently Android as my computer is broken. I tried this library on the PSP, if this is what you wanted to know.

**Comment by janisozaur** [comment]
Cool, that's good to know. Have you had any chance of running it on Windows? I'm not sure how is it handled there. I think starting with XP there is some system API for upnp support, does it make sense to use `libnatpmp` for Windows as well?

**Comment by devnoname120** [comment]
Not really, but it if it ran on the PSP, it should run smoothly on Windows as well. I am not very fond of using Microsoft libraries for this kind of job, because it can fail for whatever reason (service disabled, etc.), it might not work consistently on all Windows versions, and even worse, there are probably some unfixed exploits on Windows XP.

Maybe you should use MiniUPnP instead as it supports both UPnP and NAT-PMP, so there is a higher chance of success for a given router.

**Comment by devnoname120** [comment]
On a side note: while UPnP and NAT-PMP are fairly easy to support, it is not the most efficient means.

The best way lies in UDP hole-punching. It is basically sending a request to a server that will answer with the public port that was used by the router of the client to send the request. This port is then used by other clients on which they will send their requests. This process has been standardized by STUN, improved by STUNT, and a try-every-possibility solution was created: ICE.

You can learn more about the topic from this link: http://www.jdrosen.net/uploads/1/5/0/0/15008848/stun-turn-ice.pptx 

For this matter, one library that I heard of (but did not test it myself) is PJNATH: http://www.pjsip.org/pjnath/docs/html/index.htm

Nonetheless, I think that we should stick to UPnP/NAT-PMP, at least for now, as it is way easier to implement and will bring less headaches.
